### PR TITLE
RP-809: restore instantiation of script object for each pipeline run

### PR DESCRIPTION
**Pull request by gregcharles**
This implementation is a bit unusual as it takes a script object and uses it as a prototype to create new script objects. Eventually, I'd  prefer to restore  the implementation that created  the Pipeline with the script classes and configuration, rather than script objects.  However, this will take more intensive refactoring and testing.

### CODE REVIEW

File: script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java [hunk 1]
```diff
@@ -1,9 +1,13 @@
 package org.opentestsystem.rdw.script;
 
+import com.google.common.collect.ImmutableMap;
 import groovy.lang.Script;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.print.attribute.HashDocAttributeSet;
```
Comment (malaffoon) [inline script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java]: ?!

File: script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        when(repository.findByCodeAndVersion(PIPELINE_NAME, PIPELINE_VERSION))
+                .thenReturn(pipelineWithUserScript(scriptCode));
+
+        Pipeline pipeline = pipelineFactory.getPipeline(PIPELINE_NAME, PIPELINE_VERSION);
+
+        try {
+            pipeline.run("good input");
+        } catch (ScriptRuntimeException sre) {
+            throw (Exception)sre.getCause();
```
Comment (malaffoon) [inline script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java]: This seems a little ... indirect ... for a test. Are you testing that the sre cause is expected to be an ImportException? Perhaps use assertThatThrownBy or something similar?

File: script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+        Pipeline pipeline = pipelineFactory.getPipeline(PIPELINE_NAME, PIPELINE_VERSION);
+
+        try {
+            pipeline.run("bad input");
+            fail("Expected a script exception for bad input");
+        } catch (ScriptRuntimeException sre) {
+        }
+
+        pipeline.run("good input");
```
Comment (malaffoon) [inline script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java]: This should really be asserting something ... perhaps just `assertThat(pipeline.run("good input").asString()).isEqualTo("good input");`

File: script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        when(repository.findByCodeAndVersion(PIPELINE_NAME, PIPELINE_VERSION))
+                .thenReturn(pipelineWithUserScript(scriptCode));
+
+        Pipeline pipeline = pipelineFactory.getPipeline(PIPELINE_NAME, PIPELINE_VERSION);
+
+        try {
+            pipeline.run("bad input");
+            fail("Expected a script exception for bad input");
+        } catch (ScriptRuntimeException sre) {
```
Comment (malaffoon) [inline script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java]: Again, you could use assertThatThrownBy or something similar ... it makes the test read better by being explicit about what is expected.

File: script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java [hunk 5]
```diff
@@ -152,4 +156,27 @@ private ScriptRuntimeException extractCause(final Throwable t) {
...
 
+    /**
+     * Get an instance of the script to be run against the given input
+     *
+     * @param prototype prototype instance of the Pipeline script
+     * @param input input for the script to work on. The type is script dependent.
+     * @return an instance of a PipelineScript
+     * @throws Exception if there is an problem constructing the script
+     */
+    private PipelineScript getInstance(final PipelineScript prototype, final Object input) throws Exception {
```
Comment (malaffoon) [inline script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: It feels wrong to me that this isn't in the PipelineFactory somehow. I guess one could argue that this is the moral equivalent of a copy-constructor?

File: script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java [hunk 6]
```diff
@@ -31,15 +36,35 @@ public PipelineScript source(final PipelineScriptDefinition source) {
...
+            addBoundProperties(properties);
             for (Map.Entry<String,Object> entry : properties.entrySet()) {
                 this.setProperty(entry.getKey(), entry.getValue());
             }
         }
-
         return this;
     }
 
+    private void addBoundProperties(final Map<String, Object> properties) {
```
Comment (malaffoon) [inline script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java]: i sort of see the desire to have an immutable map but this is complex for something that is used only internally. That is, this entire method could be replaced with a single line up around line 40.

File: script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java [hunk 7]
```diff
@@ -152,4 +156,27 @@ private ScriptRuntimeException extractCause(final Throwable t) {
         return null;
     }
 
+    /**
+     * Get an instance of the script to be run against the given input
+     *
```
Comment (malaffoon) [inline script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: Some additional javadoc would be nice, perhaps
```
 * This constructs a copy of the script with the original bound properties and source
 * to guarantee we have a fresh instance with no memory of previous processing.
```

**Comment by malaffoon** [comment]
Hmm, i'm waffling on this one. It feels like we may want to do the short term "fix" to checkValid, and then leave the (techdebt) JIRA around to refactor it properly as you describe. Meh.

File: script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java (hunk 1 above)
Comment (gregcharles) [reply script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java]: Oops, imports now optimized. (It was a suggestion for Hash... that was HashMap, when finger moved toward the Enter key, but HasDocAttributeSet when it landed. IntelliJ is a merry prankster!)

File: script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java (hunk 2 above)
Comment (gregcharles) [reply script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java]: done

File: script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java (hunk 4 above)
Comment (gregcharles) [reply script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java]: done

File: script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java (hunk 6 above)
Comment (gregcharles) [reply script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java]: done

File: script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java (hunk 7 above)
Comment (gregcharles) [reply script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: Done

File: script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java (hunk 3 above)
Comment (gregcharles) [reply script/src/test/java/org/opentestsystem/rdw/script/TextPipelineTest.java]: done

File: script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java (hunk 5 above)
Comment (gregcharles) [reply script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: I'll put in a technical debt ticket to track this. My preference is for the construction to only happen in Pipeline, but that will require more extensive changes.

File: script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java [hunk 8]
```diff
@@ -152,4 +156,27 @@ private ScriptRuntimeException extractCause(final Throwable t) {
...
+     * @param input input for the script to work on. The type is script dependent.
+     * @return an instance of a PipelineScript
+     * @throws Exception if there is an problem constructing the script
+     */
+    private PipelineScript getInstance(final PipelineScript prototype, final Object input) throws Exception {
+        // TODO: this method should be updated to work with PipelineDefinition, not PipelineScript
+        final PipelineScript script = prototype.getClass().getConstructor().newInstance()
+                .bindProperties(prototype.getBoundProperties())
+                .source(prototype.getSource())
+                .bindInput(input);
```
Comment (malaffoon) [inline script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: And you have absolutely no performance concerns with this change? I know exam processing has I/O that is the long pole but this is a lot of introspection and object construction for every exam being processed.

File: script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java (hunk 1 above)
Comment (malaffoon) [inline script/src/main/java/org/opentestsystem/rdw/script/PipelineScript.java]: i'm confused, it's still there?

File: script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java (hunk 8 above)
Comment (gregcharles) [reply script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: I'm worried more with the reflection than the construction. Nothing against Spring, but I'd be surprised if there weren't several hundred objects created for each run of an exam. I suppose I could cache the constructors the first time I reflected for them. Still, I just tested 100,000 runs of reflect, construct, and bind input and that ran in 1.1 seconds. That didn't include binding other properties, but it did include printing out the instance references. (It ran so fast without the print that I was afraid it was just optimizing away the whole loop.)

Comment (gregcharles) [reply script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: `    public void performance() throws Exception {
        final String scriptCode = "enable('csv')\n" +
                "transformRows {row ->\n" +
                "   def value = row.bogus\n" +
                "}\n" +
                "outputCsv";

        final Class<? extends PipelineScript> scriptClass = compiler.compile(scriptCode);
        long start = System.currentTimeMillis();

        for (int i = 0; i < 100000; i++) {
            System.out.println(scriptClass.getConstructor().newInstance().bindInput(csv));
        }

        System.out.println(System.currentTimeMillis() - start);
    }
`

Comment (malaffoon) [reply script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: Ok, thanks for taking the time to play with that. I'll avoid premature optimization ;)

**Pull request merged**